### PR TITLE
Block: Try adding the left/right block hover areas

### LINKED ISSUE #3044: Consider hiding block ellipsis and movers until mouse is near them
From https://github.com/WordPress/gutenberg/issues/2983, but opened as a separate ticket because it can (or not) happen separately from other efforts.

The goal is to reduce the visual weight of the block. Since the movers are largely a desktop enhancement, we can hide them until the mouse is near them. You should still be able to tab into them, of course, which would also reveal them. The idea being that these features are power-user features, where it's okay that you have to "learn of their existance", and when you know where they are and how they work, they'll be available to you when you need them, but otherwise out of your way. Here's how it could work.

Hover near the edges of the block to surface the movers and the context menu:

block hover near right

block hover near left

This is similar to how Google Docs handles it:

docs comments

Thoughts?

### PR DESCRIPTION
**Pull request by youknowriad**
For unselected blocks, we only show movers the movers when we're close to them and same for the block settings menu.

**Testing instructions**

 - Add several blocks
 - Try hovering next to the left and the right area of the blocks
 - When you're close to the left the movers appear
 - When you're close to the right the settings menu appear

closes #3044

### CODE REVIEW

File: editor/components/block-list/with-hover-areas.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+ * Module constants
+ */
+const HOVER_AREA_SIZE = 120;
+
+const withHoverAreas = ( WrappedComponent ) => {
+	class WithHoverAreasComponent extends Component {
+		constructor() {
+			super( ...arguments );
+			this.state = {
+				hoverArea: null,
```
Comment (youknowriad) [inline editor/components/block-list/with-hover-areas.js]: I can easily add an `isHovered` state to this component to remove it from the `block.js` component but it seems that this state is slightly more complex than just keeping a boolean on whether the block is hovered or not. Might be good to do but I'll defer to @aduth here. I think you have more knowledge about the edge cases there.

**Comment by jasmussen** [comment]
Too cool.

It seems like it would be worth trying a few things to tweak this further. That is, always hide the mover and ellipsis, even when the block is selected. Perhaps combined with the hoverable area left and right being _wider_, so they invoke more easily, that will strike a nice balance. 

Can we also add a transition? Something alongthe lines of `transition: opacity .4s linear`, so they side UI fades in when you enter the hover area?

We want to make sure this works well on mobile also. That is, on mobile, these shouldn't fade out at all, as the side UI sits below the block in the mobile breakpoints.

**Comment by mtias** [comment]
Thanks for trying this. It's feeling rather nice to me and greatly reduces the UI weight. If adding this also for the selected state we have to make sure things can be targeted with keyboard.

**Comment by youknowriad** [comment]
Updated the PR. It works also for selected blocks, but the elements are tabbable even if hidden so it works fine using keyboard.

**Comment by jasmussen** [comment]
Here's a GIF:

hover areas

This seems to work really well for me now, very very nice. 

I feel like we should make the hoverable area to reveal the side controls bigger... It feels like it's below 20% now, but it could be between 25 and 33% the width of the whole block, I feel like.

Or is it measured in pixels right now? If yes, then the breakpoint where we switch between mobile side UI and desktop side UI is 600px, and so if we can't use percent, it might be worth trying to make it 200px left side, 200px neutral center, 200px right side. There's no reason not to go pretty big with this stuff, at least for a start, we can always pull back. 

Nice!

**Comment by youknowriad** [comment]
@jasmussen So right now the hover area is 120px and when the block is smaller than 2* 120px the hover area is 50% of the block. I can increase it to 200px or use a percentage, your choice :)

**Comment by jasmussen** [comment]
Let's try, and if to make the areas min-width: 160px, max-width: 240px. Or width: 30%. Either of those, I think, would be good to test as a start, and then we can tune as we go. Sound good?

**Comment by mtias** [comment]
I'm really liking this, let's get it in. :)

**Comment by jasmussen** [comment]
Dig it. This really lightens the cognitive burden on the UI, while still allowing powerful tools to be present. To echo Matías, we should definitely try this. And this fixes #3044. 

Also appreciate bundling the hover label with the side hover, I think that's perfectly good to try. 👍 👍 from me.

File: editor/components/block-list/breadcrumb.js [hunk 2]
```diff
@@ -34,27 +25,59 @@ function stopPropagation( event ) {
...
+		super( ...arguments );
+		this.state = {
+			isFocused: false,
+		};
+		this.onFocus = this.onFocus.bind( this );
+		this.onBlur = this.onBlur.bind( this );
+	}
+
+	onFocus( event ) {
+		this.setState( {
```
Comment (aduth) [inline editor/components/block-list/breadcrumb.js]: Why was this needed?

File: editor/components/block-list/with-hover-areas.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+		}
+
+		onMouseLeave() {
+			if ( this.state.hoverArea ) {
+				this.setState( { hoverArea: null } );
+			}
+		}
+
+		onMouseMove( event ) {
+			const { width, left, right } = this.container.getBoundingClientRect();
```
Comment (aduth) [inline editor/components/block-list/with-hover-areas.js]: >## What forces layout / reflow
>
>_[...]_ This is also called reflow or layout thrashing, and is common performance bottleneck.
>
>**Box metrics**
>
>- _[...]_ `elem.getBoundingClientRect()`

https://gist.github.com/paulirish/5d52fb081b3570c81e3a

And in a `mousemove`, it's likely being called hundreds of times 🙁

File: editor/components/block-list/with-hover-areas.js [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+			super( ...arguments );
+			this.state = {
+				hoverArea: null,
+			};
+			this.onMouseLeave = this.onMouseLeave.bind( this );
+			this.onMouseMove = this.onMouseMove.bind( this );
+		}
+
+		componentDidMount() {
+			// Disable reason: We use findDOMNode to avoid unnecessary extra dom Nodes
```
Comment (aduth) [inline editor/components/block-list/with-hover-areas.js]: Was this the main reason for using `mousemove` instead of some floating invisible elements which trigger on `mouseenter`?

Comment (youknowriad) [reply editor/components/block-list/with-hover-areas.js]: No, the main reason is that `mouseenter` doesn't trigger on areas with lower z-index, since these areas would need a lower z-index to allow writing in the editables etc...

File: editor/components/block-list/breadcrumb.js (hunk 2 above)
Comment (youknowriad) [reply editor/components/block-list/breadcrumb.js]: In the breadcrumb component at the moment no because it's not shown at all when the block is selected, but I didn't want to write the implementation depending on something not in the component itself. If tomorrow we decide to show this breadcrumb even on selected blocks, this will become necessary.

File: editor/components/block-list/with-hover-areas.js (hunk 3 above)
Comment (youknowriad) [reply editor/components/block-list/with-hover-areas.js]: I guess I can look for alternatives to `getBoundingClientRect`

File: editor/components/block-list/with-hover-areas.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
+/**
+ * WordPress dependencies
+ */
+import { Component, findDOMNode } from '@wordpress/element';
+
+const withHoverAreas = ( WrappedComponent ) => {
```
Comment (aduth) [inline editor/components/block-list/with-hover-areas.js]: Should we have used `@wordpress/element`'s `createHigherOrderComponent` ?

File: editor/components/block-list/style.scss [hunk 6]
```diff
@@ -699,3 +699,11 @@
 		padding-top: 6px;
 	}
 }
+
+.editor-block-breadcrumb {
```
Comment (aduth) [inline editor/components/block-list/style.scss]: Aside: I applied an invalid class name per standards; this should be `.editor-block-list__breadcrumb`

File: editor/components/block-list/with-hover-areas.js (hunk 3 above)
Comment (youknowriad) [reply editor/components/block-list/with-hover-areas.js]: Can't think of any alternative without browser reflow

**Comment by jasmussen** [comment]
Wanted to note that the use of a keyframe animation can cause some jumpiness. I will, in an incoming PR, refactor this to use just a `transition`, which should fix that.

File: editor/components/block-list/block.js [hunk 7]
```diff
@@ -510,23 +515,25 @@ export class BlockListBlock extends Component {
...
+						isHidden={ ! isHovered || hoverArea !== 'left' }
 					/>
 				) }
-				{ shouldShowSettingsMenu && ! showSideInserter && (
+				{ shouldRenderBlockSettings && (
 					<BlockSettingsMenu
 						uids={ [ uid ] }
 						rootUID={ rootUID }
 						renderBlockMenu={ renderBlockMenu }
+						isHidden={ ! isHovered || hoverArea !== 'right' }
```
Comment (iseulde) [inline editor/components/block-list/block.js]: @youknowriad Why does the toolbar only hide? Why not prevent rendering? If there's a reason, it's not immediately clear from this code.

Comment (iseulde) [reply editor/components/block-list/block.js]: Oh, thanks, I didn't read that far up. :)

File: editor/components/block-list/block.js [hunk 8]
```diff
@@ -510,23 +515,25 @@ export class BlockListBlock extends Component {
...
 					/>
 				) }
-				{ isHovered && <BlockBreadcrumb uid={ uid } /> }
+				{ shouldShowBreadcrumb && <BlockBreadcrumb uid={ uid } isHidden={ hoverArea !== 'left' } /> }
```
Comment (aduth) [inline editor/components/block-list/block.js]: Why do we have both a condition to render `BlockBreadcrumb` and a prop controlling whether it's visible. Should it be one or the other?
